design/0001-metallb-crd.md

```diff
@@ -0,0 +1,173 @@
+# MetalLB Custom Resource Controller 
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [Testing](#testing)
+- [Debugging and Troubleshooting](#debugging-and-troubleshooting)
+<!-- /toc -->
+
+## Summary
+Today MetalLB allows only configuration of layer2 and BGP protocols using Configmap objects, 
+which is simple from configuration perspective and doesn't require cluster-admin rights,
+however it doesn't have validation on the API server level, no support for API versioning and 
+no fine-grained permission model via RBAC.
+
+## Motivation
+Custom resource definition "CRD" takes advantage of kubernetes extensibility to deliver the automation advantages of cloud services like 
+provisioning, scaling and backup/restore.
+
+### Goals
+Configuring MetalLB layer2 and BGP using kubernetes custom resource "CRD" instead of ConfigMap.
+
+## Proposal
+Using SIG's [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) framework to build new CRD "MetalLB" object 
+and a controller which works as a translator from the user's configuration yaml to a MetalLB's configmap object.
+with this approach there will be no changes required for the MetalLB management plane.
+
+### Risks and Mitigations
+Users need to either use CRD or ConfigMap to provision load balancer instance but not both, if user
+mistakenly used both the CRD configuration will overwrite the user defined ConfigMap.
+
+## Design Details
+![Alt text](./metallb_crd.png?raw=true "MetalLB Kubernetes Controller")
```

Comment:
I have a preference for something like an ASCII diagram (e.g. https://asciiflow.com/ ) instead because it doesn't add a lot of bytes to the repo, it is easier to review and it can be easily updated after merging (even if you don't have the original source). Maybe other reviewers have other ideas about tooling/approached for this.

Reply:
this graph I did using draw.io if there is agreement on tool I will migrate to it

Reply:
Another idea is https://excalidraw.com/ which allows generating nice-looking diagrams and exporting PNG with the diagram's "source" embedded in the PNG metadata, allowing you to magically edit a diagram by opening the PNG file in Excalidraw.